libcaf_net/caf/net/transport_base.hpp

```diff
@@ -0,0 +1,182 @@
+      manager_(nullptr) {
+    // nop
+  }
+
+  // -- properties -------------------------------------------------------------
+
+  handle_type handle() const noexcept {
+    return handle_;
+  }
+
+  actor_system& system() {
+    return manager().system();
+  }
+
+  application_type& application() {
+    return next_layer_.application();
+  }
+
+  transport_type& transport() {
+    return *reinterpret_cast<transport_type*>(this);
+  }
+
+  endpoint_manager& manager() {
+    return *manager_;
+  }
+
+  // -- transport member functions ---------------------------------------------
+
+  virtual error init(endpoint_manager& parent) {
+    CAF_LOG_TRACE("");
+    manager_ = &parent;
+    auto& cfg = system().config();
+    auto max_header_bufs = get_or(cfg, "middleman.max-header-buffers",
+                                  defaults::middleman::max_header_buffers);
+    header_bufs_.reserve(max_header_bufs);
+    auto max_payload_bufs = get_or(cfg, "middleman.max-payload-buffers",
+                                   defaults::middleman::max_payload_buffers);
+    payload_bufs_.reserve(max_payload_bufs);
+    if (auto err = next_layer_.init(*this))
+      return err;
+    return none;
+  }
+
+  auto resolve(endpoint_manager&, const uri& locator, const actor& listener) {
+    CAF_LOG_TRACE(CAF_ARG(locator) << CAF_ARG(listener));
+    auto f = detail::make_overload(
+      [&](auto& layer) -> decltype(layer.resolve(*this, locator, listener)) {
+        return layer.resolve(*this, locator, listener);
+      },
+      [&](auto& layer) -> decltype(
+                         layer.resolve(*this, locator.path(), listener)) {
+        return layer.resolve(*this, locator.path(), listener);
+      });
+    f(next_layer_);
+  }
+
+  void new_proxy(endpoint_manager&, const node_id& peer, actor_id id) {
+    next_layer_.new_proxy(*this, peer, id);
+  }
+
+  void local_actor_down(endpoint_manager&, const node_id& peer, actor_id id,
+                        error reason) {
+    next_layer_.local_actor_down(*this, peer, id, std::move(reason));
+  }
+
+  void timeout(endpoint_manager&, atom_value value, uint64_t id) {
+    next_layer_.timeout(*this, value, id);
+  }
+
+  void set_timeout(uint64_t timeout_id, id_type id) {
+    next_layer_.set_timeout(timeout_id, id);
+  }
+
+  void handle_error(sec code) {
+    next_layer_.handle_error(code);
+  }
+
+  // -- (pure) virtual functions -----------------------------------------------
+
+  virtual void configure_read(receive_policy::config){
+    // nop
+  };
+
+  virtual bool handle_read_event(endpoint_manager&) = 0;
+
+  virtual bool handle_write_event(endpoint_manager& parent) = 0;
+
+  virtual void write_packet(id_type id, span<buffer_type*> buffers) = 0;
+
+  // -- buffer management ------------------------------------------------------
+
+  buffer_type next_header_buffer() {
+    return next_buffer_impl(header_bufs_);
+  }
+
+  buffer_type next_payload_buffer() {
+    return next_buffer_impl(payload_bufs_);
+  }
```

Comment:
Because we're trying to limit vector reallocations.`header_buffers` are always the same (small) size. If one of those `header_buffers` is passed as a `payload_buffer` it would have to be reallocated to fit the bigger payload size. Hence we're managing both buffer-types differently.

Reply:
Ah. I missed that you pass different arguments. Makes sense!